internal/accesstoken/tokenexchange.go

```diff
@@ -33,67 +32,35 @@ const (
 type exchangeTokenSource struct {
 	cfg            AccessTokenExchangeConfig
 	ctx            context.Context
-	mu             sync.Mutex
 	upstream       oauth2.TokenSource
-	upstreamToken  *oauth2.Token
 	exchangeConfig oauth2.Config
-	token          *oauth2.Token
 }
 
 // Token retrieves an OAuth 2.0 access token from the configured issuer using token exchange.
 // Tokens are reused as long as they are valid.
 // Upstream tokens used as the source for the exchange are reused as long as they are valid.
 func (s *exchangeTokenSource) Token() (*oauth2.Token, error) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	if s.token != nil && s.token.Valid() {
-		return s.token, nil
-	}
-
-	if err := s.refreshUpstream(); err != nil {
-		return s.token, err
-	}
-
-	if err := s.exchange(); err != nil {
-		return s.token, err
-	}
-
-	return s.token, nil
-}
-
-func (s *exchangeTokenSource) refreshUpstream() error {
-	if s.upstreamToken == nil || !s.upstreamToken.Valid() {
-		token, err := s.upstream.Token()
-		if err != nil {
-			return fmt.Errorf("%w: %w", ErrUpstreamTokenRequestFailed, err)
-		}
-
-		s.upstreamToken = token
+	upstreamToken, err := s.upstream.Token()
```

Comment:
comment on line 40 isnt quite correct - valid but still within the timeout window

Reply:
whoops, good catch. thanks